includes/compat/trait-wc-stripe-subscriptions.php

```diff
@@ -570,9 +570,21 @@ public function add_subscription_information_to_intent( $request, $order, $prepa
 			$parent_order    = wc_get_order( $parent_order_id );
 
 			if ( $parent_order ) {
-				$mandate = $parent_order->get_meta( '_stripe_mandate_id', true );
+				$mandate        = $parent_order->get_meta( '_stripe_mandate_id', true );
+				$renewal_amount = WC_Stripe_Helper::get_stripe_amount( $renewal_order->get_total() );
+
 				if ( ! empty( $mandate ) ) {
 					$request['mandate'] = $mandate;
+					return $request;
+				} elseif ( 0 !== $renewal_amount ) {
+					$request['payment_method_options']['card']['mandate_options']['amount_type']     = 'fixed';
+					$request['payment_method_options']['card']['mandate_options']['interval']        = $renewal_order->get_billing_period();
+					$request['payment_method_options']['card']['mandate_options']['interval_count']  = $renewal_order->get_billing_interval();
+					$request['payment_method_options']['card']['mandate_options']['amount']          = $renewal_amount;
+					$request['payment_method_options']['card']['mandate_options']['reference']       = $order->get_id();
+					$request['payment_method_options']['card']['mandate_options']['start_date']      = $renewal_order->get_time( 'start' );
+					$request['payment_method_options']['card']['mandate_options']['supported_types'] = [ 'india' ];
+
 					return $request;
 				}
```

Comment:
The entire point of this was to just see if the mandate existed on the parent order, so I don't think this is the right place to add this logic. My initial implementation would've been better served by extracting this into a function so the code would've looked like:

```php
// instead of if ( 1 === count( $renewasl ) ) { /* ... */

$renewals = wcs_get_subscriptions_for_renewal_order( $order );
$mandate  = $this->get_mandate_from_parent_order( $order );
if ( $mandate ) {
  $request['mandate'] = $mandate;
  return $request;
}
```

All of that to say: I don't think the fix is in the right place here 🙂

Reply:
@reykjalin thanks for your review and guideline. I have made some changes to incorporate some of the suggestions you made [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2633#pullrequestreview-1464268566) like saving the mandate in the parent order.

I am a bit confused about this suggestion. What would be the best place for these changes? :thinking: 
I added the changes here to support the following scenario-
The subscription initially has 0 amount. So when the user first purchases the sub, there is no mandate. Then the subscription value is updated and now it has a positive value to charge which requires a mandate. At this point when a renewal order comes there is no mandate in the parent but we must send these `mandate_options` with the request. As this is a renewal order, without this else section it returns from [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L587) without adding the `mandate_options`.

Reply:
All I meant to say was that these changes should exist outside the `if ( 1 === count( $renewal_orders ) )` statement, something like:

```php
// Check if mandate already exists.
if ( 1 === count( $renewal_orders ) ) {
    // ...
    if ( $parent_order ) {
        // ...
    }
}

// Make sure we handle renewal orders that don't have mandates.
if ( 0 !== $renewal_amount ) {
    // Continue your code here.
}
```


I suggest this because the section you initially added the code to is intended to only check if there exists a mandate on the parent order, and combining that with handling free trials (or $0 subscriptions) would be confusing.

***

I honestly think the best way to handle this is probably to do one of the following:

1. Check what the renewal amount will be when a $0 sub is processed, create a mandate with that correct amount, save that to the parent order, and use that for renewal payments — this would work with the current "check the parent order for a mandate" approach; or
2. Create the mandate on the first renewal order that has a non-zero amount (what you're doing now), and either:
    1. Save the mandate to the parent order so it can be used for future purchases (this would work with the current "check the parent order for a mandate" approach); or
    2. Save the mandate to the renewal order. This would require changing the logic here such that we don't only check the parent order for a mandate but also previous renewal orders.

***

And if that all wasn't complicated enough you can create subscriptions without a parent order, in which case (1) and (2i) wouldn't work, so (2ii) is probably the best approach? But that complexity should probably be handled in a separate PR.

Reply:
Thanks @reykjalin for the suggestions. I had I have implemented 2i now with a few additional checks before reusing the mandate. :tada: 
I would appreciate your re-review here in case I have missed any other corner cases (like the subscriptions without a parent order case). Let me know if you can think of any possible regression too.

Reply:
You've addressed the main point here (not merging "finding the mandate" and "handle 0 cost subscriptions") so I'll mark this as resolved. Danae has hit upon some great points in her review and I added a few other comments of my own to continue the related discussions 🙂 

Great work! 👏